### ISSUE TITLE
Cast model outputs not involved in loss to float

### ISSUE THREAD

**Issue by TidalPaladin** [TidalPaladin/dl-template#19]
`Prediction` computes probabilities directly from `logits`, which will fail when 16-bit predictions have been copied to CPU. Functions not involved in loss computation should be safe by casting 16-bit to 32-bit floats where applicable.
